Blog on fine-tuning a model with KAITO (#4484)
Azure/AKS: blog/_posts/2024-08-23-fine-tuning-language-models-with-kaito.md

Let’s experiment with [Phi-3-Mini-128K-Instruct](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct), a robust and high-performing model that’s a bit smaller in size than your average LLM. We’ll use a custom [Streamlit](https://streamlit.io/) chat UI app, connected to a Phi-3 inference service deployed by KAITO, and ask a basic question about AKS:

![Screenshot of untuned model response](/AKS/assets/images/fine-tuning-language-models-with-kaito/untuned-lm-with-kaito.png)
_Screenshot of un-tuned model response_

Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.

## KAITO can help with model fine-tuning

> KAITO supports Low-Rank Adaptation (LoRA) and Quantized LoRA (QLoRA), the [most common](https://arxiv.org/pdf/2402.16141) parameter-efficient fine-tuning techniques, which significantly lower the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 

Reviewer: links to what LORA, QLORA are could be useful.

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and may require higher performing GPUs in general. 

When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be added to new inferencing workspaces!

> After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

Reviewer: love the conversational tone here!

![Three hours later](/AKS/assets/images/fine-tuning-language-models-with-kaito/spongebob-transition-3-hours-later.jpg)

## Better results with fine-tuned workspaces

To use our new fine-tuned adapter, the inference workspace will now accept one or more adapters to customize model behavior.

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
```

> As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 

Reviewer: merges the adapters - can it be simplified further?

With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users. (Another way to increase model accuracy is Retrieval-Augmented Generation (RAG), which we touch on briefly in this [section](#Stay-tuned-to-make-models-smarter), coming soon to KAITO).

## Fine-tuning a foundation model is sometimes necessary

Let’s experiment with [Phi-3-Mini-128K-Instruct](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct), a robust and high-performing model that’s a bit smaller in size than your average LLM. We’ll use a custom [Streamlit](https://streamlit.io/) chat UI app, connected to a Phi-3 inference service deployed by KAITO, and ask a basic question about AKS:

> ![Screenshot of untuned model response](/AKS/assets/images/kaito-blog/untuned-lm-with-kaito.png)

Author: ```suggestion
![Screenshot of untuned model response](/AKS/assets/images/fine-tuning-language-models-with-kaito/untuned-lm-with-kaito.png)
```

_Screenshot of un-tuned model response_

Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.

## KAITO can help with model fine-tuning

> KAITO supports Low-Rank Adaption (LoRA) and Quantized LoRA (QLoRA), the [most common](https://arxiv.org/pdf/2402.16141) parameter-efficient fine-tuning techniques, which significantly lower the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 

Reviewer 2: "significantly lowers"

Author: Both techniques lower the GPU requirement for tuning, so I kept it plural

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-tuning-phi-3
resource:
  instanceType: Standard_NC24ads_A100_v4
  labelSelector:
    matchLabels:
      app: tuning-phi-3
tuning:
  preset:
    name: phi-3-mini-128k-instruct
  method: qlora
  input:
    urls:
      - https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true
  output:
    image: myregistry.azurecr.io/adapters/myadapter:0.0.1
    imagePushSecret: myregistrysecret
```

The tuning property specifies the minimum information required for starting a tuning job, including:

**Preset**: indicates the target model for tuning (KAITO uses preset configurations to generate all Kubernetes sources needed to run the Job). 

**Method**: users can choose LoRA or QLoRA (quantized LoRA, for lower GPU memory usage) as the tuning method. 

**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).

**Output**: where the adapter is stored, as a container image or any other storage type supported by Kubernetes.

> > The choice of GPU SKU is critical since model fine-tuning normally requires more GPU memory compared to model inference. To avoid GPU Out-Of-Memory, NVIDIA A100 or higher tier GPUs are recommended. 

Reviewer 2: Is there ever an option to map LLM's to GPU's that they "fit into" for finetuning.  A/H100 are very expensive and potentially overkill for a Phi-3 for example

Reviewer 2: This is not a blocker by the way, just a "wouldn't it be nice"

Author: Good question - for inferencing with Phi-3 just the NC6 series GPU is needed, but KAITO currently needs A100 or higher to fine-tune in a reasonable amount of time (3 hours or less), and in the future more optimizations will be made to lower this requirement.

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and may require higher performing GPUs in general. 

When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be added to new inferencing workspaces!

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

> ![Three hours later](/AKS/assets/images/kaito-blog/spongebob-transition-3-hours-later.jpg)

Author: ```suggestion
![Three hours later](/AKS/assets/images/fine-tuning-language-models-with-kaito/spongebob-transition-3-hours-later.jpg)
```

## Better results with fine-tuned workspaces

To use our new fine-tuned adapter, the inference workspace will now accept one or more adapters to customize model behavior.

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
```

As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 

Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:

> ![Screenshot of tuned model response](/AKS/assets/images/kaito-blog/fine-tuned-lm-with-kaito.png)

Author: ```suggestion
![Screenshot of tuned model response](/AKS/assets/images/fine-tuning-language-models-with-kaito/fine-tuned-lm-with-kaito.png)
```

---
title: "Fine tune language models with KAITO on AKS"
description: "Make your AI model even smarter, in just a few steps"
date: 2024-08-23
author: Sachi Desai
categories:
- ai
- add-ons
tags:
- kaito
---

## Introduction

You may have heard of the [Kubernetes AI Toolchain Operator](https://github.com/Azure/kaito/tree/main) (KAITO) announced at Ignite 2023 and KubeCon Europe this year. The open source project has gained popularity in recent months by introducing a streamlined approach to AI model deployment and flexible infrastructure provisioning on Kubernetes.

> With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users. (Another way to increase model accuracy is Retrieval-Augmented Generation (RAG), which we touch on briefly in this [section](#"Stay tuned"-to-make-models-smarter), coming soon to KAITO).

Author: ```suggestion
With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users. (Another way to increase model accuracy is Retrieval-Augmented Generation (RAG), which we touch on briefly in this [section](#"Stay"-"tuned"-to-make-models-smarter), coming soon to KAITO).
```

---
title: "Fine tune language models with KAITO on AKS"
description: "Make your AI model even smarter, in just a few steps"
date: 2024-08-23
author: Sachi Desai
categories:
- ai
- add-ons
tags:
- kaito
---

## Introduction

You may have heard of the [Kubernetes AI Toolchain Operator](https://github.com/Azure/kaito/tree/main) (KAITO) announced at Ignite 2023 and KubeCon Europe this year. The open source project has gained popularity in recent months by introducing a streamlined approach to AI model deployment and flexible infrastructure provisioning on Kubernetes.

> With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users.

Author: ```suggestion
With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users. (Another way to increase model accuracy is Retrieval-Augmented Generation (RAG), which we touch on briefly in this [section](#"Stay tuned"-to-make-models-smarter)).
```

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and may require higher performing GPUs in general. 

When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be added to new inferencing workspaces!

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

> ![Three hours later](https://i.ytimg.com/vi/Qn_PfS8ZsBg/maxresdefault.jpg)

Author: ```suggestion
![Three hours later](/AKS/assets/images/kaito-blog/spongebob-transition-3-hours-later.jpg)
```

> > This particular fine-tuning job took a little over 3 hours, but job completion duration will vary depending on the foundation model and the size of the input dataset.

Author: ```suggestion
```

> 

Author: ```suggestion
```
